di.go

```diff
@@ -347,117 +347,129 @@ func (s *DIServer[T]) diDone(ctx context.Context, msg io.Reader) (struct{}, erro
 		CertChain: &certChain,
 		Entries:   nil,
 	}
-	if err := s.maybeAutoExtend(ctx, ov); err != nil {
-		return struct{}{}, fmt.Errorf("error extending voucher: %w", err)
+	if s.BeforeVoucherPersist != nil {
+		if err := s.BeforeVoucherPersist(ctx, ov); err != nil {
+			return struct{}{}, fmt.Errorf("error in callback before new voucher is persisted: %w", err)
+		}
 	}
 	if err := s.Vouchers.NewVoucher(ctx, ov); err != nil {
 		return struct{}{}, fmt.Errorf("error storing voucher: %w", err)
 	}
-	if err := s.maybeAutoTO0(ctx, ov); err != nil {
-		return struct{}{}, fmt.Errorf("error auto-registering device for rendezvous: %w", err)
+	if s.AfterVoucherPersist != nil {
+		if err := s.AfterVoucherPersist(ctx, ov); err != nil {
+			return struct{}{}, fmt.Errorf("error in callback after new voucher is persisted: %w", err)
+		}
 	}
 	return struct{}{}, nil
 }
 
-func (s *DIServer[T]) maybeAutoExtend(ctx context.Context, ov *Voucher) error {
-	if s.AutoExtend == nil {
-		return nil
-	}
-
-	mfgKey := ov.Header.Val.ManufacturerKey
-	keyType, rsaBits := mfgKey.Type, mfgKey.RsaBits()
-	owner, _, err := s.AutoExtend.ManufacturerKey(ctx, keyType, rsaBits)
-	if err != nil {
-		return fmt.Errorf("error getting %s manufacturer key: %w", keyType, err)
-	}
-	nextOwner, _, err := s.AutoExtend.OwnerKey(ctx, keyType, rsaBits)
-	if err != nil {
-		return fmt.Errorf("error getting %s owner key: %w", keyType, err)
-	}
-	switch owner.Public().(type) {
-	case *ecdsa.PublicKey:
-		nextOwner, ok := nextOwner.Public().(*ecdsa.PublicKey)
-		if !ok {
-			return fmt.Errorf("owner key must be %s", keyType)
-		}
-		extended, err := ExtendVoucher(ov, owner, nextOwner, nil)
+// AutoExtend extends a voucher and replaces the value pointed to with the
+// newly extended voucher.
+//
+// This function is meant to be used as a callback in DIServer.
+func AutoExtend(
+	manufacturerKey func(ctx context.Context, keyType protocol.KeyType, rsaBits int) (crypto.Signer, []*x509.Certificate, error),
+	ownerKey func(ctx context.Context, keyType protocol.KeyType, rsaBits int) (crypto.Signer, []*x509.Certificate, error),
+) func(context.Context, *Voucher) error {
```

Comment:
Not loving this signature, but interfaces don't fit that well either. Thoughts? Same for AutoTO0.